gherkin/php/src/GherkinDocumentBuilder.php

```diff
@@ -0,0 +1,371 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Cucumber\Gherkin;
+
+use Cucumber\Gherkin\Parser\Builder;
+use Cucumber\Gherkin\Parser\RuleType;
+use Cucumber\Gherkin\Parser\TokenType;
+use Cucumber\Gherkin\ParserException\AstBuilderException;
+use Cucumber\Messages\Background;
+use Cucumber\Messages\Comment;
+use Cucumber\Messages\DataTable;
+use Cucumber\Messages\DocString;
+use Cucumber\Messages\Examples;
+use Cucumber\Messages\Feature;
+use Cucumber\Messages\FeatureChild;
+use Cucumber\Messages\GherkinDocument;
+use Cucumber\Messages\Id\IdGenerator;
+use Cucumber\Messages\Location as MessageLocation;
+use Cucumber\Messages\Rule;
+use Cucumber\Messages\RuleChild;
+use Cucumber\Messages\Scenario;
+use Cucumber\Messages\Step;
+use Cucumber\Messages\TableCell;
+use Cucumber\Messages\TableRow;
+use Cucumber\Messages\Tag;
+use LogicException;
+
+/**
+ * @implements Builder<GherkinDocument>
+ */
+final class GherkinDocumentBuilder implements Builder
+{
+    /** @var list<Comment> */
+    private array $comments = [];
+
+    /** @var non-empty-list<AstNode> */
+    private array $stack;
+
+    public function __construct(
+        private readonly string $uri,
+        private readonly IdGenerator $idGenerator
+    ) {
+        $this->reset($uri);
+    }
+
+    public function build(Token $token): void
+    {
+        if (is_null($token->matchedType)) {
+            throw new \LogicException('Token was not yet matched');
+        }
+
+        $ruleType = RuleType::cast($token->matchedType);
+
+        if ($token->matchedType == TokenType::Comment) {
+            $this->comments[] = new Comment($this->getLocation($token, 0), $token->matchedText ?? '');
+        } else {
+            $this->currentNode()->add($ruleType, $token);
+        }
+    }
+
+    public function startRule(RuleType $ruleType): void
+    {
+        array_push($this->stack, new AstNode($ruleType));
+    }
+
+    public function endRule(RuleType $ruleType): void
+    {
+        $node = array_pop($this->stack);
+        if ($transformedNode = $this->getTransformedNode($node)) {
+            $this->currentNode()->add($node->ruleType, $transformedNode);
+        }
+    }
+
+    public function getResult(): GherkinDocument
+    {
+        $document = $this->currentNode()->getSingle(GherkinDocument::class, Ruletype::GherkinDocument);
+
+        if (is_null($document)) {
+            throw new \LogicException('GherkinDocument was not built from source, but no parse errors');
+        }
+
+        return $document;
+    }
+
+    public function reset(string $uri): void
+    {
+        $this->stack = [new AstNode(RuleType::None)];
+    }
+
+    private function currentNode(): AstNode
+    {
+        return $this->stack[array_key_last($this->stack)];
+    }
+
+    /**
+     * @return object|string|list<object>|null
+     */
+    private function getTransformedNode(AstNode $node): object|string|array|null
```

Comment:
At 189 lines. Would it be terrible to pull all the internals of each case, out into a named, possibly private methods? Not asking for the change, but trying to understand why you did not?

Was it the rule of three (these each, presumably only being used here?)

Reply:
This is a good point, I've split it out consequently can use a `match`